File: src/content/docs/topics/cs/coding-problems/bit-manipulation/136-single-number-approach2.go
package main

import (
	"fmt"
	"sort"
)

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func singleNumber(nums []int) int {
	sort.Ints(nums)                             // L1: O(n log n)
	for i := 0; i < len(nums)-1; i += 2 {      // L2: scan in steps of 2
		if nums[i] != nums[i+1] {              // L3: O(1)
			return nums[i]
		}
	}
	return nums[len(nums)-1]                    // L4: last element is single
}

func main() {
	assert(singleNumber([]int{2, 2, 1}) == 1)
	assert(singleNumber([]int{4, 1, 2, 1, 2}) == 4)
	assert(singleNumber([]int{1}) == 1)
	assert(singleNumber([]int{0, 0, 99}) == 99)
	assert(singleNumber([]int{-1, -1, 42}) == 42)
	fmt.Println("all tests pass")
}

File: src/content/docs/topics/cs/coding-problems/bit-manipulation/136-single-number-approach2.py
def single_number(nums: list[int]) -> int:
    nums.sort()                             # L1: O(n log n)
    for i in range(0, len(nums) - 1, 2):   # L2: scan in steps of 2
        if nums[i] != nums[i + 1]:         # L3: O(1)
            return nums[i]
    return nums[-1]                         # L4: last element is single

assert single_number([2, 2, 1]) == 1
assert single_number([4, 1, 2, 1, 2]) == 4
assert single_number([1]) == 1
assert single_number([0, 0, 99]) == 99
assert single_number([-1, -1, 42]) == 42
assert single_number([2 ** 31 - 1]) == 2 ** 31 - 1
print("all tests pass")
